### PR TITLE
Enable optional FP32 compute for LM Head

### PR DESCRIPTION
**Pull request by narutolhy**
## Motivation

Large-vocab matmul at the final projection (“LM Head”) can benefit from FP32 for numerical stability and HF parity (e.g., logprob alignment), while we still want the forward path (RMSNorm/Attn/MLP) to stay in BF16/FP16 for performance and CUDA Graph stability.
This PR adds an opt-in flag to compute only the LM Head in FP32, without perturbing earlier layers or embedding dtypes, and without disabling CUDA Graph elsewhere.

## Modifications

1) CLI & configuration plumbing
Added a new CLI flag --lm-head-fp32 (backward-compatible alias --enable-lm-head-fp32) to compute only the LM Head in FP32.
2) LogitsProcessor reads the flag
In LogitsProcessor.__init__, persist the flag:
`self.use_lm_head_fp32 = global_server_args_dict["lm_head_fp32"]`
3) Guarded FP32 compute in _get_logits
When enabled (--lm-head-fp32), Switch to FP32 only for the final projection:
`if self.use_lm_head_fp32:
    logits = torch.matmul(
        hidden_states.to(torch.float32), lm_head.weight.to(torch.float32).T
    )`
4) Unit tests
Added test/srt/test_lm_head_fp32.py (CUDA):

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by narutolhy** [comment]
@zhaochenyang20 please take a look, thanks

File: test/srt/test_lm_head_fp32.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+class TestLMHeadFP32(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        if not torch.cuda.is_available():
+            raise unittest.SkipTest("needs CUDA GPU")
+
+    def _make_lp(self, vocab_size, enable_fp32):
+        global_server_args_dict["enable_dp_lm_head"] = False
+        global_server_args_dict["enable_lm_head_fp32"] = enable_fp32
```
[the remark this answers is not in the document]
Comment (zhaochenyang20) [reply test/srt/test_lm_head_fp32.py]: Shall refer to other parameters. But I belive SGLang has a lot of parameters named as `enable_xxx`.

File: python/sglang/srt/layers/logits_processor.py [hunk 2]
```diff
@@ -461,21 +462,34 @@ def _get_logits(
...
+                    logits = torch.ops.sgl_kernel.weight_packed_linear(
+                        hidden_states.to(lm_head.weight.dtype),
+                        lm_head.weight,
+                        None,  # bias
+                        True,  # is_vnni
+                    )
+                else:
+                    logits = torch.matmul(
+                        hidden_states.to(lm_head.weight.dtype), lm_head.weight.T
+                    )
```
[the remark this answers is not in the document]
Comment (zhaochenyang20) [reply python/sglang/srt/layers/logits_processor.py]: I think this is better:

```python
if self.use_lm_head_fp32:
    # Use full float32 precision for higher accuracy
    logits = torch.matmul(
        hidden_states.to(torch.float32), lm_head.weight.to(torch.float32).T
    )
elif use_intel_amx_backend(lm_head):
    # Use Intel AMX-optimized backend for performance
    logits = torch.ops.sgl_kernel.weight_packed_linear(
        hidden_states.to(lm_head.weight.dtype),
        lm_head.weight,
        None,  # bias
        True,  # is_vnni
    )
else:
    # Default behavior, use the original data type
    logits = torch.matmul(
        hidden_states.to(lm_head.weight.dtype), lm_head.weight.T
    )
```

File: python/sglang/srt/server_args.py [hunk 3]
```diff
@@ -1280,6 +1281,11 @@ def add_cli_args(parser: argparse.ArgumentParser):
             choices=["auto", "fp8_e5m2", "fp8_e4m3"],
             help='Data type for kv cache storage. "auto" will use model data type. "fp8_e5m2" and "fp8_e4m3" is supported for CUDA 11.8+.',
         )
+        parser.add_argument(
+            "--lm-head-fp32",
+            action="store_true",
+            help="If set, the LM head outputs (logits) are in FP32.",
+        )
```
[the remark this answers is not in the document]
Comment (zhaochenyang20) [reply python/sglang/srt/server_args.py:1288]: reasonable sugguestions

File: test/srt/test_lm_head_fp32.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    def test_flag_true_fp16_activations(self):
+        a, b = self._run_case(torch.float16, True, torch.float32)
+        self.assertEqual(a, torch.float32)
+        self.assertEqual(b, torch.float32)
+
+    def test_flag_true_bf16_activations(self):
+        a, b = self._run_case(torch.bfloat16, True, torch.float32)
+        self.assertEqual(a, torch.float32)
+        self.assertEqual(b, torch.float32)
```
[the remark this answers is not in the document]
Comment (zhaochenyang20) [reply test/srt/test_lm_head_fp32.py]: I am not sure if this is reasonable.

File: docs/advanced_features/server_arguments.md [hunk 5]
```diff
@@ -95,6 +95,7 @@ Please consult the documentation below and [server_args.py](https://github.com/s
 | `--quantization` | The quantization method. | None |
 | `--quantization-param-path` | Path to the JSON file containing the KV cache scaling factors. This should generally be supplied, when KV cache dtype is FP8. Otherwise, KV cache scaling factors default to 1.0, which may cause accuracy issues. | None |
 | `--kv-cache-dtype` | Data type for kv cache storage. 'auto' will use model data type. 'fp8_e5m2' and 'fp8_e4m3' is supported for CUDA 11.8+. | auto |
+| `--lm-head-fp32` | If set, the LM head outputs (logits) are in FP32. | False |
```
Comment (zhaochenyang20) [inline docs/advanced_features/server_arguments.md]:
```python
| `--lm-head-fp32` | If set, the LM head outputs (logits) are in FP32, which gives better precision on log probs for RL. | False |
```

File: python/sglang/srt/layers/logits_processor.py [hunk 6]
```diff
@@ -220,6 +220,7 @@ def __init__(
         self.config = config
         self.logit_scale = logit_scale
         self.use_attn_tp_group = global_server_args_dict["enable_dp_lm_head"]
+        self.use_lm_head_fp32 = global_server_args_dict["lm_head_fp32"]
```
Comment (zhaochenyang20) [inline python/sglang/srt/layers/logits_processor.py]: name it as `use_fp32_lm_head`?

File: test/srt/test_lm_head_fp32.py
Comment (zhaochenyang20) [inline test/srt/test_lm_head_fp32.py:1]: move this file to `test/srt/rl`?

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: test/srt/test_lm_head_fp32.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    def compute_dp_attention_metadata(self): ...
+
+
+class TestLMHeadFP32(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        if not torch.cuda.is_available():
+            raise unittest.SkipTest("needs CUDA GPU")
+
+    def _make_lp(self, vocab_size, enable_fp32):
```
Comment (zhaochenyang20) [inline test/srt/test_lm_head_fp32.py]: what does `lp` stand for? Could we use a longer naming?

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: test/srt/test_lm_head_fp32.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        global_server_args_dict["enable_dp_lm_head"] = False
+        global_server_args_dict["enable_lm_head_fp32"] = enable_fp32
+        cfg = SimpleNamespace(vocab_size=vocab_size, final_logit_softcapping=None)
+        return LogitsProcessor(cfg, skip_all_gather=True, logit_scale=None)
+
+    def _run_case(self, hs_dtype, enable_fp32, w_dtype):
```
Comment (zhaochenyang20) [inline test/srt/test_lm_head_fp32.py]: could we use full naming, like `hidden_state_dtype, weights_dtype`. ALso, def this function as:

```python
def _run_case(self, hidden_state_dtype, weights_dtype, enable_fp32):
```

File: test/srt/test_lm_head_fp32.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        device = "cuda"
+        B, H, V = 2, 64, 128
+        hs = torch.randn(B, H, dtype=hs_dtype, device=device)
+        head = LMHeadStub(V, H, dtype=w_dtype, device=device)
+        meta = DummyMeta()
+        lp = self._make_lp(V, enable_fp32)
+
+        orig_mm = torch.matmul
+        orig_li = F.linear
```
Comment (zhaochenyang20) [inline test/srt/test_lm_head_fp32.py]: I tends to use longer naming for these 😂

File: test/srt/test_lm_head_fp32.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+        state = {"called": False, "op": None, "a": None, "b": None}
```
Comment (zhaochenyang20) [inline test/srt/test_lm_head_fp32.py]: could we name `op` to `probe_target`?

File: test/srt/test_lm_head_fp32.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+            if not state["called"]:
+                state.update(called=True, op="linear", a=x.dtype, b=w.dtype)
+            return orig_li(x, w, bias)
+
+        with patch("torch.matmul", new=probe_matmul), patch(
+            "torch.nn.functional.linear", new=probe_linear
+        ):
+            logits = lp._get_logits(hs, head, meta)
+
+        self.assertEqual(logits.dtype, torch.float32)
```
Comment (zhaochenyang20) [inline test/srt/test_lm_head_fp32.py]: we also tested `false`, so this may not always be true?

File: test/srt/test_lm_head_fp32.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+    def _make_lp(self, vocab_size, enable_fp32):
+        global_server_args_dict["enable_dp_lm_head"] = False
+        global_server_args_dict["enable_lm_head_fp32"] = enable_fp32
+        cfg = SimpleNamespace(vocab_size=vocab_size, final_logit_softcapping=None)
+        return LogitsProcessor(cfg, skip_all_gather=True, logit_scale=None)
+
+    def _run_case(self, hs_dtype, enable_fp32, w_dtype):
+        device = "cuda"
+        B, H, V = 2, 64, 128
```
Comment (zhaochenyang20) [inline test/srt/test_lm_head_fp32.py]: `BATCH_SIZE, HIDDEN_SIZE, VOCAB_SIZE`

File: test/srt/test_lm_head_fp32.py (hunk 10 above)
Comment (zhaochenyang20) [reply test/srt/test_lm_head_fp32.py]: Adding comments for the four states in the `state`. What's the key and value stand for.

File: test/srt/test_lm_head_fp32.py
Comment (zhaochenyang20) [inline test/srt/test_lm_head_fp32.py:1]: Why not also check `a, b` in the `_run_case`, rather than seperate it?

```python
def _run_case(self, hs_dtype, enable_fp32, w_dtype, expected_a_dtype, expected_b_dtype):
        device = "cuda"
        B, H, V = 2, 64, 128
        hs = torch.randn(B, H, dtype=hs_dtype, device=device)
        head = LMHeadStub(V, H, dtype=w_dtype, device=device)
        meta = DummyMeta()
        lp = self._make_lp(V, enable_fp32)

        orig_mm = torch.matmul
        orig_li = F.linear

        state = {"called": False, "op": None, "a": None, "b": None}

        def probe_matmul(a, b, *args, **kw):
            if not state["called"]:
                state.update(called=True, op="matmul", a=a.dtype, b=b.dtype)
            return orig_mm(a, b, *args, **kw)

        def probe_linear(x, w, bias=None):
            if not state["called"]:
                state.update(called=True, op="linear", a=x.dtype, b=w.dtype)
            return orig_li(x, w, bias)

        with patch("torch.matmul", new=probe_matmul), patch(
            "torch.nn.functional.linear", new=probe_linear
        ):
            logits = lp._get_logits(hs, head, meta)

        # 所有的断言都在这里进行
        self.assertEqual(logits.dtype, torch.float32)
        self.assertEqual(hs.dtype, hs_dtype)
        self.assertTrue(state["called"], "no call lm head matlmul/linear")
        self.assertEqual(state["a"], expected_a_dtype)
        self.assertEqual(state["b"], expected_b_dtype)

    def test_flag_true_fp16_activations(self):
        self._run_case(torch.float16, True, torch.float32, torch.float32, torch.float32)

    def test_flag_true_bf16_activations(self):
        self._run_case(torch.bfloat16, True, torch.float32, torch.float32, torch.float32)

    def test_flag_false_fp16_path(self):
        self._run_case(torch.float16, False, torch.float16, torch.float16, torch.float16)

    def test_flag_false_bf16_path(self):
        self._run_case(torch.bfloat16, False, torch.bfloat16, torch.bfloat16, torch.bfloat16)
```

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Review by zhaochenyang20 — APPROVED** [review]
great job

**Review by zhaochenyang20 — APPROVED** [review]
Great job

**Review by merrymercy — APPROVED** [review]
